qiskit/transpiler/passes/routing/lookahead_swap.py

```diff
@@ -303,43 +305,67 @@ def _map_free_gates(layout, gates, coupling_map):
+        max_gates = 50 + 10 * len(state.coupling_map.physical_qubits)
+
+    layout_map = state.layout._v2p
+    out = 0
+    for gate in gates[:max_gates]:
+        if not gate["partition"]:
+            continue
+        qubits = gate["partition"][0]
+        if len(qubits) == 2:
+            out += state.coupling_map.distance(layout_map[qubits[0]], layout_map[qubits[1]])
+    return out
 
-    return sum(
-        coupling_map.distance(*(layout[q] for q in gate["partition"][0]))
-        for gate in gates[:max_gates]
-        if gate["partition"] and len(gate["partition"][0]) == 2
-    )
+
+def _score_state_with_swap(swap, state, gates):
+    """Calculate the relative score for a given SWAP.
+
+    Returns:
+        float: the score of the given swap.
+        Tuple[int, int]: the input swap that should be performed.
+        _SystemState: an updated system state with the new layout contained.
+    """
+    trial_layout = state.layout.copy()
+    trial_layout.swap(*swap)
+    new_state = state._replace(layout=trial_layout)
+    return _calc_layout_distance(gates, new_state), swap, new_state
 
 
 def _score_step(step):
     """Count the mapped two-qubit gates, less the number of added SWAPs."""
     # Each added swap will add 3 ops to gates_mapped, so subtract 3.
-    return len([g for g in step["gates_mapped"] if len(g.qargs) == 2]) - 3 * len(
-        step["swaps_added"]
-    )
+    return len([g for g in step.gates_mapped if len(g.qargs) == 2]) - 3 * len(step.swaps_added)
 
 
-def _transform_gate_for_layout(gate, layout):
+def _transform_gate_for_system(gate, state):
     """Return op implementing a virtual gate on given layout."""
-    mapped_op_node = deepcopy([n for n in gate["graph"].nodes() if isinstance(n, DAGOpNode)][0])
+    mapped_op_node = copy.copy(_first_op_node(gate["graph"]))
 
-    device_qreg = QuantumRegister(len(layout.get_physical_bits()), "q")
-    mapped_qargs = [device_qreg[layout[a]] for a in mapped_op_node.qargs]
+    device_qreg = state.register
+    layout_map = state.layout._v2p
+    mapped_qargs = [device_qreg[layout_map[a]] for a in mapped_op_node.qargs]
     mapped_op_node.qargs = mapped_qargs
 
     return mapped_op_node
 
 
-def _swap_ops_from_edge(edge, layout):
+def _swap_ops_from_edge(edge, state):
     """Generate list of ops to implement a SWAP gate along a coupling edge."""
-    device_qreg = QuantumRegister(len(layout.get_physical_bits()), "q")
+    device_qreg = state.register
     qreg_edge = [device_qreg[i] for i in edge]
 
     # TODO shouldn't be making other nodes not by the DAG!!
     return [DAGOpNode(op=SwapGate(), qargs=qreg_edge, cargs=[])]
+
+
+def _first_op_node(dag):
+    """Get the first op node from a DAG."""
+    # This doesn't use `DAGCircuit.op_nodes` because that function always consumes the entire
+    # iterator to create a list, whereas we only need the first element.
+    return next(node for node in dag.nodes() if isinstance(node, DAGOpNode))
```

Comment:
Reading through the code I don't think this matters but this is the first op node based on insertion order not necessarily topological order was that your intent or did you want the first node from a topological ordering of the dag?

Reply:
Ok, that's what it looked like to me too but I wanted to confirm. The code just read a little weird to me because the "first" node could actually be something in the middle of the dag or at the end as it's dependent on insertion order.